File: frontend/src/api/giftSolution.ts
import { API_BASE, type ProductCardData } from './chat';

export type GiftShape = 'single_gift' | 'gift_combo' | 'either';

export interface GiftShapeDecisionData {
  shape: GiftShape;
  confidence: number;
  reason: string;
  signals: string[];
  recommended_product_count: number;
  use_combo_optimizer: boolean;
}

export interface GiftSolutionResponse {
  solution_id: string;
  shape_decision: GiftShapeDecisionData;
  title: string;
  summary: string;
  products: ProductCardData[];
  total_amount?: number | string | null;
  recommendation_reason: string;
  giving_timing: string;
  giving_place: string;
  gift_talk: string;
  recipient_reaction_reply: string;
  packaging_advice: string;
  avoid_tips: string[];
  follow_up_question?: string | null;
  selected_plan_type?: string | null;
  plan_judge_reason?: string | null;
}

export async function generateGiftSolution(payload: {
  message: string;
  conversation_id?: string | null;
  user_id?: string | null;
  recommendation_strategy?: 'llm_direct' | 'hybrid_algorithm' | 'llm_rerank' | null;
  allow_generic_recommendation?: boolean;
  use_profile?: boolean;
}) {
  const response = await fetch(`${API_BASE}/gift-solution/generate`, {
    method: 'POST',
    headers: {
      'content-type': 'application/json',
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    throw new Error(`Generate gift solution failed: ${response.status}`);
  }

  return response.json() as Promise<GiftSolutionResponse>;
}

File: backend/app/schemas/gift_solution.py
from __future__ import annotations

from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.product import ProductCard
from app.schemas.recommendation import RecommendationStrategy


GiftShape = Literal["single_gift", "gift_combo", "either"]


class GiftShapeDecision(BaseModel):
    shape: GiftShape
    confidence: float = 0.0
    reason: str
    signals: list[str] = Field(default_factory=list)
    recommended_product_count: int = 1
    use_combo_optimizer: bool = False


class GiftSolutionGenerateRequest(BaseModel):
    message: str
    user_id: str | None = None
    conversation_id: str | None = None
    recommendation_strategy: RecommendationStrategy | None = "hybrid_algorithm"
    allow_generic_recommendation: bool = True
    use_profile: bool = True


class GiftSolutionResponse(BaseModel):
    solution_id: str
    shape_decision: GiftShapeDecision
    title: str
    summary: str
    products: list[ProductCard]
    total_amount: Decimal | None = None
    recommendation_reason: str
    giving_timing: str
    giving_place: str
    gift_talk: str
    recipient_reaction_reply: str
    packaging_advice: str
    avoid_tips: list[str] = Field(default_factory=list)
    follow_up_question: str | None = None
    selected_plan_type: str | None = None
    plan_judge_reason: str | None = None
    pipeline: dict[str, object] = Field(default_factory=dict)
